xarray/core/types.py

```diff
@@ -156,8 +156,15 @@ def copy(
 
 T_DataArrayOrSet = TypeVar("T_DataArrayOrSet", bound=Union["Dataset", "DataArray"])
 
-# Maybe we rename this to T_Data or something less Fortran-y?
+# Use `T_Xarray` for most functions, when the return type matches the input type — for
```

Comment:
This is true for any TypeVar.

The main difference between `T_DataArrayOrSet` and `T_Xarray` is:
- `T_DataArrayOrSet` has to be used when the input cannot be resolved to either `Dataset` or `DataArray` but is a `Union[Dataset, DataArray]`
- `T_Xarray` has to be used when you have multiple inputs that can be either `Dataset` or `DataArray` but all of them must be the same.

Reply:
> `T_Xarray` has to be used when you have multiple inputs that can be either `Dataset` or `DataArray` but all of them must be the same.

Yes, but surely another reason is that the return type of passing a `DataArray` to:

```
def gen(bar: T_Xarray) -> T_Xarray:
    return bar
```

...will be typed as a `DataArray`. 

But the return type of passing a `DataArray` to:

```
def union(bar: T_DataArrayOrSet) -> T_DataArrayOrSet:
    return bar
```

...will be typed as a `Union[DataArray, Dataset]`

...and so `union(ds).to_array()` will fail? (`.to_array` is a methodonly on a `Dataset`)

Reply:
`reveal_type(union(xr.Dataset()))`
>Revealed type is "xarray.core.dataset.Dataset""

Only if you pass a Union to this function will the return type also be a Union.

Reply:
OK, thanks a lot, and sorry for my misunderstanding. (I've really been out of this for a while...)

Let me take another pass at the comments so we can distill this down.

Reply:
No worries :)

Reply:
You have entered a quite tricky part of type checking. It took me probably days to figure out when to use which types that we have in xarray (I also have to admit that the sheer abundance of mixin classes does not make it easy).